Re-write permissions if running as root

SimonRichardson opened the pull request:
The following commit attempts to detect if you're executing with
sudo and to assign the lock file with the correct permissions, to
prevent cryptic error messages.

The code detects sudo using environment variables and correctly
parses the UID and GID from those values, before chown'ing the file.

If at any point there is an error whilst changing the owner, it
will continue as root and on the next run, it will give information
about the file in question. This should then aid diagnosing the
problem without changing the underlying existing flow.

CC: @howbazaar
Background information: https://bugs.launchpad.net/juju/+bug/1758369

SimonRichardson committed:
Re-write permissions if running as root

The following commit attempts to detect if you're executing with
sudo and to assign the lock file with the correct permissions, to
prevent cryptic error messages.

The code detects sudo using environment variables and correctly
parses the UID and GID from those values, before chown'ing the file.

If at any point there is an error whilst changing the owner, it
will continue as root and on the next run, it will give information
about the file in question. This should then aid diagnosing the
problem without changing the underlying existing flow.

hmlanigan commented on mutex_flock.go:
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
+			}
+			return
+		}
+		// Attemptting to open a lock file as root whilst using sudo can cause
```
nit: attempting

hmlanigan commented on mutex_flock.go:
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
+			return
+		}
+		// Attemptting to open a lock file as root whilst using sudo can cause
+		// the lock file to have the wrong permissions. Subsequent calls to
```
the wrong permissions for a non-sudo user....

hmlanigan reviewed (commented):
just a couple of nits, i realized i don't know enough about how sudo and the SUDO_* env vars work to know if this is a good approach.  :-/

SimonRichardson committed:
PR comments

Changing comment to make more sense and also spelling

howbazaar commented on mutex_flock.go:
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
 	}
 
 	flockName := filepath.Join(os.TempDir(), "juju-"+name)
+	chownFromRoot := func() (err error) {
```
You don't need the named error response.

howbazaar commented on mutex_flock.go:
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
+			var uid, gid int
+			uid, err = strconv.Atoi(os.Getenv("SUDO_UID"))
+			if err != nil {
+				err = errors.Annotate(err, "parsing SUDO_UID")
```
Even with named return values, I prefer the explicit approach.

SimonRichardson replied:
Fine by me, I'll get on it!

howbazaar commented on mutex_flock.go:
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
+		}
+		return nil
+	}
+	open := func() (fd int, err error) {
```
Can we not use named return values and just have explicit returns? There are only two.

SimonRichardson replied:
Of course

howbazaar commented on mutex_test.go:
```diff
@@ -284,6 +287,27 @@ func (s *mutexSuite) TestMutexNotInherited(c *gc.C) {
 }
 
+// TestFilePermissions tests that the file permissions are correct.
+func (s *mutexSuite) TestFilePermissions(c *gc.C) {
```
While kinda useful, this test doesn't actually test the code you have written at all because it will never be run with sudo.

SimonRichardson replied:
Yeah, I'll make a complimentary test function, good spot.

SimonRichardson replied:
fixed in https://github.com/juju/mutex/pull/6/commits/b5c51c193415eabe150a62058ec2985c57395433

howbazaar reviewed (approved):
I think the approach is fine, but I'd like the functions not to use named return values.

SimonRichardson committed:
Test acquire with sudo envars

The following commit introduces the ability to mock out the sudo envars
so that we can test the acquire as sudo path through the code.

We introduce the mocking framework here so that we can improve the
dependency testing of the acquire and also eventually improve the
quality of the tests using expectations.

SimonRichardson committed:
Enable race checking by default

Although superfical, it's actually easy to introduce racey issues
into a library, so we should make an effort to always check from
the outset.

SimonRichardson committed:
Don't change method signatures

We don't need to change the method signatures here, let's just depend
on the local var to begin with.

SimonRichardson committed:
Standardise library

Standardise the mutex library to have the same look and feel as
other packages.

SimonRichardson committed:
Test merge jobs

SimonRichardson commented:
$$merge$$
